File: Software/WebUi/calibration.php
<?php
/*
 * AquaPi - sterownik akwariowy oparty o Raspberry Pi
 *
 * Copyright (C) 2012-2020 AquaPi Developers
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License Version 2 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
 * USA.
 *
 */
 
include("init.php");
	$records = 0;
	if($_GET['id']) {
		$records = $db->GetOne('SELECT COUNT(*) FROM calib_data WHERE `calib_data_interface_id` = ?',array($_GET['id']));
	}
	$smarty->assign('calibration_records', $records);
	
	if($_GET['action'] == 'edit') {
		$smarty->assign('edit', $_GET['id']);	
		// pobierz dane kalibracyjne z bazy, a jeśli ich nie ma, to przyjmij jakieś dane domyślne
		if ($records == 0) {
			// nie ma danych kalibracyjnych
			$smarty->assign('raw1', -100);
			$smarty->assign('cal1', -100);
			$smarty->assign('raw2', 100);
			$smarty->assign('cal2', 100);
		} else {
			// są dane kalibracyjne, pobierz je wyślij do przeglądarki
			$calib_data	= $db->GetRow('SELECT * FROM calib_data WHERE `calib_data_interface_id` = ?',array($_GET['id']));
			$smarty->assign('raw1', $calib_data['calib_data_raw1']);
			$smarty->assign('cal1', $calib_data['calib_data_cal1']);
			$smarty->assign('raw2', $calib_data['calib_data_raw2']);
			$smarty->assign('cal2', $calib_data['calib_data_cal2']);
		}		
	} elseif($_GET['action'] == 'delete') {
		// usuń dane kalibracyjne z bazy (o ile są)
		if ($records == 0) {
			// nie ma czego kasować
			$SESSION->redirect("calibration.php?action=edit&id=".$_GET['id']);
		} else {
			// jest co kasować, to wykonajmy query
			$db->Execute('DELETE FROM calib_data WHERE `calib_data_interface_id` = ?',array($_GET['id']));
			ReloadDaemonConfig();
			$SESSION->redirect("calibration.php?action=edit&id=".$_GET['id']);
		}
	} elseif($_GET['action'] == 'update') {
		if ($records == 0) {
			// nie ma jeszcze rekordów kalibracyjnych więc insert
			$db->Execute('INSERT INTO `calib_data` (`calib_data_interface_id`, `calib_data_raw1`, `calib_data_cal1`, `calib_data_raw2`, `calib_data_cal2`)
							VALUES (?,?,?,?,?)',array($_GET['id'],$_POST['raw1'],$_POST['cal1'],$_POST['raw2'],$_POST['cal2']));
		} else {
			// są dane kalibracyjne, więc update
			$db->Execute('UPDATE `calib_data` SET `calib_data_raw1` = ?, `calib_data_cal1`=?, `calib_data_raw2`=?, `calib_data_cal2`=? 
							WHERE `calib_data_interface_id`=?',array($_POST['raw1'],$_POST['cal1'],$_POST['raw2'],$_POST['cal2'],$_GET['id']));			
		}	
		ReloadDaemonConfig();
		$SESSION->redirect("calibration.php?action=edit&id=".$_GET['id']);
	}

$daemon_data 	= @simplexml_load_string(IPC_CommandWithReply("status"));
$interfaces = GetInterfaces();

$smarty->assign('daemon_data', $daemon_data);
$smarty->assign('interfaces', $interfaces);
$smarty->assign('title', 'Kalibracja czujników wejściowych');
$smarty->display('calibration.tpl');
?>

File: Software/daemon/calibration.c
/*
 * AquaPi - sterownik akwariowy oparty o Raspberry Pi
 *
 * Copyright (C) 2012-2020 AquaPi Developers
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License Version 2 as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307,
 * USA.
 *
 */

#include <stdbool.h>

struct _calib_data {
	int id;
	int interface_id;
	double data_raw1;
	double data_cal1;
	double data_raw2;
	double data_cal2;
	double a;
	double b;

} calib_data[100];

int calib_data_count;	
 
void Calibration_ReadSettings() {
	MYSQL_RES *result;
	MYSQL_ROW row;

	// Wczytanie danych kalibracyjnych
	Log("Wczytanie danych kalibracyjnych",E_DEV);
	calib_data_count = 	-1;
	DB_Query("SELECT calib_data_id,calib_data_interface_id,calib_data_raw1,calib_data_cal1,calib_data_raw2,calib_data_cal2 FROM calib_data");
	result = mysql_store_result(conn);
	while ((row = mysql_fetch_row(result))) {
		calib_data_count++;
		calib_data[calib_data_count].id = atof(row[0]);
		calib_data[calib_data_count].interface_id = atof(row[1]);
		calib_data[calib_data_count].data_raw1 = atof(row[2]);
		calib_data[calib_data_count].data_cal1 = atof(row[3]);
		calib_data[calib_data_count].data_raw2 = atof(row[4]);
		calib_data[calib_data_count].data_cal2 = atof(row[5]);
		// wylicz współczynniki funkcji liniowej - a i b
		if (calib_data[calib_data_count].data_raw1 == calib_data[calib_data_count].data_raw2 || calib_data[calib_data_count].data_cal1 == calib_data[calib_data_count].data_cal2) {
			// jeśli raw1 = raw2 albo cal1 = cal2 to współczynników nie da się wyliczyć, więc przyjmujemy brak korekcji
			calib_data[calib_data_count].a = 1;
			calib_data[calib_data_count].b = 0;
		} else {
			calib_data[calib_data_count].a = (calib_data[calib_data_count].data_cal2-calib_data[calib_data_count].data_cal1)/(calib_data[calib_data_count].data_raw2-calib_data[calib_data_count].data_raw1);
			calib_data[calib_data_count].b = calib_data[calib_data_count].data_cal1-calib_data[calib_data_count].a*calib_data[calib_data_count].data_raw1;
		}			
	}	
	mysql_free_result(result);
}

void Calibration_Process() {		
	int x,y,calibrated;
	
	// przejedźmy się po interfejsach, i ew. dokonaj na nich korekcji
	for (y=0; y <= interfaces_count; y++) {
		calibrated = false;
		for (x=0; x <= calib_data_count; x++) {
			if (calib_data[x].interface_id == interfaces[y].id) {
				interfaces[y].measured_value = calib_data[x].a * interfaces[y].raw_measured_value + calib_data[x].b;
				calibrated = true;
			}
		}
		if (!calibrated) {
			interfaces[y].measured_value = interfaces[y].raw_measured_value;
		}
	}
}
	
void Calibration_Debug() {	
	char buff[200];
	int x;
	
	Log("═════════════════════ Zrzut danych kalibracyjnych ═════════════════",E_DEV);	
	Log("┌─────┬───────┬───────┬───────┬───────┬───────┬──────────── Dane ──",E_DEV);
	Log("│Id   │Interf.│Raw1   │Cal1   │Raw2   │Cal2   │A      │B      │",E_DEV);
	Log("├─────┼───────┼───────┼───────┼───────┼───────┼───────┼───────┤",E_DEV);				
	for(x = 0; x <= calib_data_count; x++) {
		sprintf(buff,"│%5i│%7i│%7.2f│%7.2f│%7.2f│%7.2f│%7.2f│%7.2f│",calib_data[x].id,calib_data[x].interface_id,calib_data[x].data_raw1,calib_data[x].data_cal1,calib_data[x].data_raw2,calib_data[x].data_cal2,calib_data[x].a,calib_data[x].b); 
		Log(buff,E_DEV);
	}
	Log("└─────┴───────┴───────┴───────┴───────┴───────┴───────┴───────┘",E_DEV);
}

